DataManagementSystem/Client/ConsistencyChecks.py

```diff
@@ -0,0 +1,836 @@
+  """
+  def __init__( self, interactive = True, transClient = None, dm = None, fc = None ):
+    """ c'tor
+
+        One object for every production/BkQuery/directoriesList...
+    """
+    self.interactive = interactive
+    self.transClient = TransformationClient() if transClient is None else transClient
+    self.dm = DataManager() if dm is None else dm
+    self.fc = FileCatalog() if fc is None else fc
+
+    self.dirac = Dirac()
+
+    # Base elements from which to start the consistency checks
+    self._prod = 0
+    self._bkQuery = None
+    self._fileType = []
+    self._fileTypesExcluded = []
+    self._lfns = []
+    self.noLFC = False
+    self.directories = []
+
+    # Accessory elements
+    self.runsList = []
+    self.runStatus = None
+    self.fromProd = None
+    self.transType = ''
+    self.cachedReplicas = {}
+
+    self.prcdWithDesc = []
+    self.prcdWithoutDesc = []
+    self.prcdWithMultDesc = []
+    self.nonPrcdWithDesc = []
+    self.nonPrcdWithoutDesc = []
+    self.nonPrcdWithMultDesc = []
+    self.descForPrcdLFNs = []
+    self.descForNonPrcdLFNs = []
+    self.removedFiles = []
+
+    self.absentLFNsInFC = []
+    self.existLFNsNoSE = {}
+    self.existLFNsBadReplicas = {}
+    self.existLFNsBadFiles = {}
+
+    self.commonAncestors = {}
+    self.multipleDescendants = {}
+    self.ancestors = {}
+
+  ################################################################################
+
+  def getReplicasPresence( self, lfns ):
+    """ get the replicas using the standard DataManager.getReplicas()
+    """
+    present = set()
+    notPresent = set()
+
+    chunkSize = 1000
+    printProgress = ( len( lfns ) > chunkSize )
+    startTime = time.time()
+    self.__write( "Checking replicas for %d files%s" %
+                  ( len( lfns ), ( ' (chunks of %d)' % chunkSize ) if printProgress else '... ' ) )
+    for chunk in breakListIntoChunks( lfns, chunkSize ):
+      if printProgress:
+        self.__write( '.' )
+      for _ in range( 1, 10 ):
+        res = self.dm.getReplicas( chunk )
+        if res['OK']:
+          present.update( res['Value']['Successful'] )
+          self.cachedReplicas.update( res['Value']['Successful'] )
+          notPresent.update( res['Value']['Failed'] )
+          break
+        else:
+          gLogger.error( "\nError getting replicas from FC, retry", res['Message'] )
+          time.sleep( 0.1 )
+    self.__write( ' (%.1f seconds)\n' % ( time.time() - startTime ) )
+
+    gLogger.info( "Found %d files with replicas and %d without" % ( len( present ), len( notPresent ) ) )
+    return list( present ), list( notPresent )
+
+  ################################################################################
+
+  def getReplicasPresenceFromDirectoryScan( self, lfns ):
+    """ Get replicas scanning the directories. Might be faster.
+    """
+
+    dirs = {}
+    present = []
+    notPresent = []
+    compare = True
+
+    for lfn in lfns:
+      dirN = os.path.dirname( lfn )
```

Comment:
Yes, I know, but then you are doing this test
+      if lfn == dirN + '/':
+        compare = False

which I thought could be more nicely achieved using realpath. But I did not see that you actually use dirN after, so forget what I said

Btw, be careful, if you want to play it safe, you need to make sure that if a directory is given as param, it finishes with a '/', otherwise you will add the parent directory to dirs